Performance.md

```diff
@@ -51,6 +57,79 @@ Type relationships between interfaces are also cached, as opposed to intersectio
 + }
 ```
 
+## Using Type Annotations
+
+Adding type annotations, especially return types, can save the compiler a lot of work.
+In part, this is because named types tend to be more compact than anonymous types (which the compiler might infer), which reduces the amount of time spend reading and writing declaration files (e.g. for incremental builds).
+Type inference is very convenient, so there's no need to do this universally - however, it can be a useful thing to try if you've identified a slow section of your code.
+
+```diff
+- import { otherFunc } from "other";
+-
+- export function func() {
+-     return otherFunc();
+- }
++ import { otherFunc, otherType } from "other";
++
++ export function func(): otherType {
++     return otherFunc();
++ }
+```
+
+## Preferring Base Types Over Unions
+
+Union types are great - they let you express the range of possible values for a type.
+
+```ts
+interface WeekdaySchedule {
+  day: "Monday" | "Tuesday" | "Wednesday" | "Thursday" | "Friday";
+  wake: Time;
+  startWork: Time;
+  endWork: Time;
+  sleep: Time;
+}
+
+interface WeekendSchedule {
+  day: "Saturday" | "Sunday";
+  wake: Time;
+  familyMeal: Time;
+  sleep: Time;
+}
+
+declare function printSchedule(schedule: WeekdaySchedule | WeekendSchedule);
+```
+
+However, they also come with a cost.
+Every time an object is assigned to a `Schedule`, it has to be compared to each element of the union.
+For a two-element union, this is trivial and inexpensive.
+However, if your union has more than a dozen elements, it can cause real problems.
+For instance, to eliminate redundant members from a union, the elements have to be compared pairwise, which is quadratic.
+Similarly, if you intersect large unions, distribution can result in enormous types.
+One way to avoid this is to use subtypes, rather than unions.
+
+```ts
+interface Schedule {
+  day: "Monday" | "Tuesday" | "Wednesday" | "Thursday" | "Friday" | "Saturday" | "Sunday";
+  wake: Time;
+  sleep: Time;
+}
+
+interface WeekdaySchedule extends Schedule {
+  day: "Monday" | "Tuesday" | "Wednesday" | "Thursday" | "Friday";
+  startWork: Time;
+  endWork: Time;
+}
+
+interface WeekendSchedule extends Schedule {
+  day: "Saturday" | "Sunday";
+  familyMeal: Time;
+}
+
+declare function printSchedule(schedule: Schedule);
+```
+
+More realistically, consider declaring a base `HtmlElement` type, which `DivElement`, `ImgElement`, etc extend, rather than using a union `DivElement | /*...*/ | ImgElement | /*...*/`.
```

Comment:
```suggestion
A more realistic example of this might come up when trying to model every built-in DOM element type.
In this case, it would be preferable to create a base `HtmlElement` type with common members, which `DivElement`, `ImgElement`, etc. extend, rather than to create the exhaustive union `DivElement | /*...*/ | ImgElement | /*...*/`.
```